volttron/platform/install_agents.py

```diff
@@ -239,95 +202,117 @@ def install_agent_directory(opts):
+    wheel = open(path, 'rb')
+    _log.debug(f"Connecting to {peer} to install {path}")
+    channel = connection.vip.channel(peer, 'agent_sender')
 
-    opts.connection.server.core.stop()
+    def send():
+        nonlocal wheel, channel
+        sha512 = hashlib.sha512()
+        try:
+            # TODO: RMQ channel???
+            # Note sending and receiving through a channel all communication
+            # is binary for zmq (RMQ may be different for this functionality)
+            #
+            
+            first = True
+            op = None
+            size = None
+            while True:
+                if first:
+                    first = False
+                    # Wait for peer to open compliment channel
+                    resp = jsonapi.loadb(channel.recv())
+                    _log.debug(f"Got first response {resp}")
+
+                    if len(resp) > 1:
+                        op, size = resp
+                    else:
+                        op = resp[0]
+                    
+                    if op != 'fetch':
+                        raise ValueError(f'First channel response must be fetch but was {fetch}')
+
+                if op == 'fetch':
+                    chunk = wheel.read(size)
+                    if chunk:
+                        _log.debug(f"Op was fetch sending {size}")
+                        sha512.update(chunk)
+                        channel.send(chunk)
+                    else:
+                        _log.debug(f"Op was fetch sending complete")
+                        channel.send(b'complete')
+                        gevent.sleep(10)
+                        break
+                elif op == 'checksum':
+                    _log.debug(f"sending checksum {sha512.hexdigest()}")
+                    channel.send(sha512.digest())
 
-    # This is where we need to exit so the script doesn't continue after installation.
-    sys.exit(0)
+                _log.debug("Waiting for next response")
+                # wait for next response
+                resp = jsonapi.loadb(channel.recv())
 
+                if len(resp) > 1:
+                    op, size = resp
+                else:
+                    op = resp[0]
+
+        finally:
+            _log.debug("Closing wheel and channel.")
+            wheel.close()
+            channel.close(linger=0)
+            del channel
+
+    _log.debug(f"calling install_agent on {peer} using channel {channel.name}")
+    result = connection.vip.rpc.call(
+        peer, 'install_agent', os.path.basename(path), channel.name,
+        vip_identity, publickey, secretkey, force)
+    
+    task = gevent.spawn(send)
+    result.rawlink(lambda glt: task.kill(block=False))
+    _log.debug("Completed sending of agent across.")
+    gevent.wait([result])
+    _log.debug(f"After wait result is {result}")
+    return result
+
+
+def send_agent(opts, wheel_file, vip_identity, publickey, secretkey, force):
```

Comment:
Could accept just opts.connection or opts.connection.server as the first parameter

Reply:
good point fixed